ipv6_ext: fixed extension header handling

Yonezawa-T2 opened the pull request:
`gnrc_ipv6_ext_demux` was incorrect, results in ICMPv6 checksum error. This patch fixes them. I confimed that RIOT does not raise ICMPv6 checksum errors now.


```diff
-    ext = ((ipv6_ext_t *)(((uint8_t *)pkt->data) + sizeof(ipv6_hdr_t)));
+    ext = (ipv6_ext_t *) pkt->data;
```

IPv6 header is already marked by `gnrc_ipv6.c`.

```diff
              default:
                 c = false;
-                offset += ((ext->len * IPV6_EXT_LEN_UNIT) + IPV6_EXT_LEN_UNIT);
-                ext = ipv6_ext_get_next((ipv6_ext_t *)ext); 
```

This incorrectly skips ICMPv6 header by 8 bytes.


```diff
-    ext_snip = gnrc_pktbuf_mark(pkt, offset, GNRC_NETTYPE_IPV6);
+    ext_snip = gnrc_pktbuf_mark(pkt, offset, GNRC_NETTYPE_IPV6_EXT); 
```

`gnrc_icmpv6_demux` searches for `GNRC_NETTYPE_IPV6` to compute the pseudo header, but this is an extension header rather than a IPv6 header, results in a checksum error. I added new `GNRC_NETTYPE_IPV6_EXT` type but `GNRC_NETTYPE_UNDEF` may be sufficient.

```diff
-  USEMODULE += ipv6_ext 
+  # USEMODULE += ipv6_ext
```

`USEMODULE += ipv6_ext` results in the following error.

```
...
"make" -C ~/Documents/native_workspace/RIOT/sys/net/network_layer/ipv6/ext
make[3]: *** No targets specified and no makefile found.  Stop.
```

I'm not sure just commenting out is appropriate but it works. If it is appropriate, I will delete it completely.

cgundogan commented on sys/net/gnrc/network_layer/ipv6/ext/gnrc_ipv6_ext.c:
```diff
@@ -20,6 +20,9 @@
 
 #include "net/gnrc/ipv6/ext.h"
 
+#define ENABLE_DEBUG    (1)
```
`ENABLE_DEBUG` should be `0`

cgundogan commented:
>...
"make" -C ~/Documents/native_workspace/RIOT/sys/net/network_layer/ipv6/ext
make[3]: *** No targets specified and no makefile found.  Stop.
I'm not sure just commenting out is appropriate but it works. If it is appropriate, I will delete it completely.

There is no `Makefile` in `RIOT/sys/net/network_layer/ipv6/ext` so this error message is expected. It should work if you build for an application (e.g. `gnrc_networking`) ?

Yonezawa-T2 commented:
@cgundogan 
The message is produced when I build an application with `gnrc_ipv6_ext` enabled. When `gnrc_ipv6_ext` is enabled in Makefile of the application, `ipv6_ext` is also enabled by Makefile.dep.

Yonezawa-T2 commented:
Makefile.dep

```
ifneq (,$(filter gnrc_ipv6_ext,$(USEMODULE)))
  USEMODULE += ipv6_ext
  USEMODULE += gnrc_ipv6
endif
```

sys/Makefile

```
ifneq (,$(filter ipv6_ext,$(USEMODULE)))
    DIRS += net/network_layer/ipv6/ext
endif
```

sys/Makefile is updated by #3594.

authmillenon commented on Makefile.dep:
```diff
@@ -211,7 +211,7 @@ ifneq (,$(filter ipv6_ext_rh,$(USEMODULE)))
 
 ifneq (,$(filter gnrc_ipv6_ext,$(USEMODULE)))
-  USEMODULE += ipv6_ext
+  # USEMODULE += ipv6_ext
```
What is the reasoning for this?

authmillenon replied:
Sorry, found it in your description ^^"

authmillenon replied:
Please just remove this line or introduce `ipv6_ext` as a pseudo module. I'm open to both :-).

Yonezawa-T2 commented:
Removed `USEMODULE += ipv6_ext` completely.

authmillenon commented on sys/net/gnrc/network_layer/ipv6/gnrc_ipv6.c:
```diff
@@ -126,6 +126,8 @@ void gnrc_ipv6_demux(kernel_pid_t iface, gnrc_pktsnip_t *pkt, uint8_t nh)
                 return;
             }
+            /* gnrc_ipv6_ext_demux calls gnrc_ipv6_demux, so stop processing here */
+            return;
```
This will keep external subscribers to any of those extension (e.g. raw sockets) from getting these packets, too. They are send in https://github.com/RIOT-OS/RIOT/pull/4654/files#diff-74976d45a42eaa063e3561b5fd36f615R141 ff.

Use `break` instead.

Yonezawa-T2 commented:
Updated to notify all extension headers. A concern is that they are notified backward. The payload is notified first, then extension headers are notified after that.

Yonezawa-T2 commented:
Subscribers for extension headers are passed payload, not extension header. I will fix this.

Yonezawa-T2 commented:
Updated.

Why does `gnrc_pktbuf_mark` set `users` of newly created snip to 1, not to `pkt->users`? When `pkt->users > 1` and `pkt` is released, the newly created snip is freed while `pkt` is not. So I copied `pkt->users` to newly created snip. Do I missed something?


Here is transitions of the reference counts of pktbuf:

initial state in `gnrc_ipv6_demux`:
  `pkt`: 1

after `gnrc_pktbuf_hold` just before `gnrc_ipv6_ext_demux`:
  `pkt`: 2

after `gnrc_pktbuf_mark` in `gnrc_ipv6_ext_demux`:
  `pkt->next` == `ext_snip`: 1
  `pkt`: 2

after `ext_snip->users = pkt->users` in `gnrc_ipv6_ext_demux`:
  `pkt->next` == `ext_snip`: 2
  `pkt`: 2

before `_dispatch_rcv_pkt` in recursivelly called `gnrc_ipv6_demux`:
  `pkt->next` (== `ext_snip` in `gnrc_ipv6_ext_demux`): 2 + `receiver_num`
  `original_pkt` == `pkt`: 2 + `receiver_num`

after `_dispatch_rcv_pkt`:
  `pkt->next`: 2
  `original_pkt` == `pkt`: 2

after `gnrc_pktbuf_release(original_pkt)`:
  `pkt->next`: 1
  `original_pkt` == `pkt`: 1

before `_dispatch_rcv_pkt` in original `gnrc_ipv6_demux`:
  `pkt` == `ext_snip`: 1 + `receiver_num`
  `original_pkt`: 1

after `_dispatch_rcv_pkt`:
  `pkt` == `ext_snip`: 1
  `original_pkt`: 1

after `gnrc_pktbuf_release(original_pkt)`:
  `pkt` == `ext_snip`: 0
  `original_pkt`: 0

Yonezawa-T2 commented:
Another concern: do we need `return` after `_decapsulate` when `nh` is `PROTNUM_IPV6` to prevent double notification and double free? What if both extension headers and encapsulated IPv6 packet are exist?

Taking decapsulation into account, we should notify extension headers before processing payload. Alternatively, `gnrc_pktbuf_remove_snip` need consider `users`. `_decapsulate` also need to remove extension headers as well as outer IPv6 header.

authmillenon commented:
> Updated to notify all extension headers. A concern is that they are notified backward. The payload is notified first, then extension headers are notified after that.

This is intentional!  See [doc on packets for that](http://doc.riot-os.org/group__net__gnrc__pkt.html#details). 

> Why does gnrc_pktbuf_mark set users of newly created snip to 1, not to pkt->users? When pkt->users > 1 and pkt is released, the newly created snip is freed while pkt is not. So I copied pkt->users to newly created snip. Do I missed something?

Because there is only one thread (the creator) holding this particular snip. The users counter is intended to be snip local, so that copying single headers is easy (also see packet doc for that).

authmillenon commented:
The rest I'll comment on, when I've got a little more overview.

Yonezawa-T2 commented:
> This is intentional! See doc on packets for that. 

The document seems not state notification order. Normally, notification is made from lower layer to higher layer. The link layer dispatch `GNRC_NETTYPE_SIXLOWPAN`, then 6LoWPAN dispatch `GNRC_NETTYPE_IPV6`, then IPv6 dispatch `GNRC_NETTYPE_ICMPV6`, `GNRC_NETTYPE_UDP`, or other payload type. So I thought dispatching extension header before payload is natural.

> The users counter is intended to be snip local

If I understand it correctly, `gnrc_pktbuf_hold` increments couters of given node and its ancestors in pktbuf tree. `gnrc_pktbuf_release` decrements given node and its ancestors as well. So `users` must be equal to or greater than the sum of `users` of its children. `gnrc_pktbuf_mark` inserts a new node between given node and its parent, so `users` of the new node must be equal to the `users` of the old node.

Yonezawa-T2 commented:
It seems that in general, calling `gnrc_pktbuf_start_write` before `gnrc_pktbuf_mark` is the proper manner, but if we call `gnrc_pktbuf_start_write` in `gnrc_ipv6_ext_demux`, IPv6 payload is copied for each extension header since `gnrc_ipv6_demux` hold the packet to notify extension headers later. If we dispatch notification for extension headers before processing payload, we don't need to hold the packet, so that `gnrc_pktbuf_start_write` will not copy the packet.

Yonezawa-T2 commented:
Dispatching extension headers before payload. Two `switch`es are ugly, but I have no other idea.

Yonezawa-T2 commented:
Rebased.

Yonezawa-T2 commented:
Reminder:
This branch will probably conflict with #4774. Update this once #4774 is merged.

Yonezawa-T2 commented:
Squashed, rebased, and resolved conflicts with #4774.

Updated a lot to make it works. I also made a sample application that sends a packet to itself with extension headers.

https://github.com/Yonezawa-T2/RIOT/tree/fix_ipv6_ext_test

Run `gnrc_networking` example which should show following debug output:

```
ipv6: Received (src = fd01::1, dst = fd01::2, next header = 0, length = 40)
ipv6: forward nh = 0 to other threads
ipv6: unable to forward packet as no one is interested in it
ipv6: handle extension header (nh = 0)
ipv6: forward nh = 43 to other threads
ipv6: unable to forward packet as no one is interested in it
ipv6: handle extension header (nh = 43)
ipv6: waiting for incoming message.
ipv6: GNRC_NETAPI_MSG_TYPE_RCV received
ipv6: Received (src = fd01::1, dst = fd01::3, next header = 0, length = 40)
ipv6: forward nh = 0 to other threads
ipv6: unable to forward packet as no one is interested in it
ipv6: handle extension header (nh = 0)
ipv6: forward nh = 43 to other threads
ipv6: unable to forward packet as no one is interested in it
ipv6: handle extension header (nh = 43)
ipv6: waiting for incoming message.
ipv6: GNRC_NETAPI_MSG_TYPE_RCV received
ipv6: Received (src = fd01::1, dst = fd01::2, next header = 0, length = 40)
ipv6: forward nh = 0 to other threads
ipv6: unable to forward packet as no one is interested in it
ipv6: handle extension header (nh = 0)
ipv6: forward nh = 43 to other threads
ipv6: unable to forward packet as no one is interested in it
ipv6: handle extension header (nh = 43)
ipv6: forward nh = 17 to other threads
ipv6: Send receive command for 0x1060b8 to 5
ipv6: waiting for incoming message.
pkt->users: 0
```

It configures the network interface with addresses `fd01::02` and `fd01::03`. Then it sends a packet to `fd01::02` with a routing extension header containing addresses `fd01::03` and `fd01::02`. So the packet should be forwarded from `fd01::02` to `fd01::03`, then to `fd01::02` again.

The packet has a Hop-by-Hop extension header that should be ignored.

The application also asserts that the packet is released.

cgundogan commented:
@Yonezawa-T2 looks great so far, will test thoroughly later. Can you include your `test application` as a `test application` in `tests`? Would be nice to have this residing there for all eternity..

Yonezawa-T2 commented:
Added the test application.

cgundogan commented:
oops, commented on commit level.

cgundogan commented on sys/include/net/gnrc/ipv6.h:
```diff
@@ -117,11 +117,22 @@ kernel_pid_t gnrc_ipv6_init(void);
+ *                     v           v
+ * IPv6 <- IPv6_EXT <- IPv6_EXT <- UNDEF
+ *
+ * This situation may happen when the packet has a sournce routing extention
```
/sournce/source/

cgundogan replied:
/extention/extension/

cgundogan commented on sys/include/net/gnrc/ipv6/ext.h:
```diff
@@ -40,16 +40,26 @@ extern "C" {
- * @param[in] iface The receiving interface.
- * @param[in] pkt   A packet.
- * @param[in] nh    A protocol number (see @ref net_protnum).
+ * This situation may happen when the packet has a sournce routing extention
```
/sournce/source/,/extention/extension/

authmillenon commented on sys/net/gnrc/network_layer/ipv6/ext/gnrc_ipv6_ext.c:
```diff
@@ -23,81 +23,271 @@
+ * @return The duplicated snip, if succeeded.
+ * @return NULL, if no space is left in the packet buffer.
+ */
+static gnrc_pktsnip_t *_duplicate_upto(gnrc_pktsnip_t *pkt, gnrc_nettype_t type)
```
This function might be interesting to be put to the `pkt` module in general (If it's not needed its optimized out anyway ;-)).

Yonezawa-T2 replied:
Promoted them as `gnrc_pkt_len_upto` and `gnrc_pktbuf_duplicate_upto`.

authmillenon commented on tests/gnrc_ipv6_ext/README.md:
```diff
@@ -0,0 +1,43 @@
+ipv6: forward nh = 17 to other threads
+ipv6: Send receive command for 0x1060b8 to 5
+ipv6: waiting for incoming message.
+pkt->users: 0
```
Can you write a pexpect script for that? With @kaspar030's recent addition of a testsuite in #4965 this should be a breeze.

Yonezawa-T2 replied:
Added pexpect script. Rebased to incorporate #4965.

authmillenon replied:
Just realized, that you need `ENABLE_DEBUG` in `gnrc_ipv6.c` set to 1 for this. Guess this is one of this cases, where an automated way to set these would really come in handy :D.

authmillenon replied:
(not scope of this PR though).

authmillenon commented on sys/include/net/gnrc/nettype.h:
```diff
@@ -63,6 +63,9 @@ typedef enum {
     GNRC_NETTYPE_IPV6,          /**< Protocol is IPv6 */
 #endif
+#ifdef MODULE_GNRC_IPV6_EXT
+    GNRC_NETTYPE_IPV6_EXT,      /**< Protocol is IPv6 extension header */
```
While I see the need for this `nettype`, I am a bit weary about how this will mess-up the `netreg`. Do we allow for registration to this? And if yes are the demux contexts the same as the next header numbers? Is registration to `GNRC_NETTYPE_IPV6` with the next header number of an extension header still possible?

Yonezawa-T2 replied:
`gnrc_ipv6_demux`:
```
_dispatch_rcv_pkt(current->type, GNRC_NETREG_DEMUX_CTX_ALL, current);
_dispatch_rcv_pkt(GNRC_NETTYPE_IPV6, nh, current);
```

For example, if `nh == 43` (IPv6-Route), it dispatches

```
_dispatch_rcv_pkt(GNRC_NETTYPE_IPV6_EXT, GNRC_NETREG_DEMUX_CTX_ALL, current);
_dispatch_rcv_pkt(GNRC_NETTYPE_IPV6, 43, current);
```

In addition, someone may want to dispatch `GNRC_NETTYPE_IPV6_EXT` with the next header:

```
_dispatch_rcv_pkt(GNRC_NETTYPE_IPV6_EXT, 43, current);
```

I'm not sure who actually want this.

authmillenon commented:
Except of the `nettype_t` considerations (no code change necessary, just some reasoning), I'm fine with the changes GNRC-wise. As for actual usage I reassign to @cgundogan as he already worked with extension headers.

authmillenon commented:
What's the progress on this?

Yonezawa-T2 commented:
rebased

DipSwitch commented:
@cgundogan Is it ok for @Yonezawa-T2 to squash?

cgundogan commented:
sure, @Yonezawa-T2 please squash. At first sight the code looks good, but I didn't do a thorough testing, yet. @authmillenon did you ran the tests?

authmillenon commented on tests/gnrc_ipv6_ext/main.c:
```diff
@@ -0,0 +1,125 @@
+#define MAIN_QUEUE_SIZE     (8)
+static msg_t _main_msg_queue[MAIN_QUEUE_SIZE];
+
+static void _send_packet() {
```
`s/()/(void)/`

Yonezawa-T2 replied:
fixed

authmillenon commented:
> @authmillenon did you ran the tests?

not until now. There seems to be a timeout on native. I thought you used this to implement RPL non-storing mode, too.

cgundogan commented:
>not until now. There seems to be a timeout on native. I thought you used this to implement RPL non-storing mode, too.

nope, there's no non-storing mode as of yet. I only implemented the source-routing-header (RH4) and did not run the ipv6 extension header internals of RIOT.

I will run the pexpect tests for this PR later ..

Yonezawa-T2 commented:
squashed

cgundogan commented:
>not until now. There seems to be a timeout on native. I thought you used this to implement RPL non-storing mode, too.

timeout? When I ran `make test` it terminated successfully. Could you elaborate on that timeout?

authmillenon commented on tests/gnrc_ipv6_ext/README.md:
```diff
@@ -0,0 +1,43 @@
+
+This test sends a packet to itself with extension headers. This is based on gnrc_networking example.
+
+Enable debug output of `gnrc_ipv6.c` before run. When the test is run, it should show the following debug output:
```
> Enable debug output of `gnrc_ipv6.c` before run.

might there be a way to not require that for the test? Because this does not allow for easy automation.

cgundogan replied:
Can this be done in a follow-up? Let's get this merged finally.

authmillenon replied:
Yeah.

cgundogan commented:
ACK, but awaiting @authmillenon answer

authmillenon commented:
Just for clarification:

> timeout? When I ran make test it terminated successfully. Could you elaborate on that timeout?

I ran into that timeout because I did not enabled debug in `gnrc_ipv6.c` ;-)

cgundogan commented:
@Yonezawa-T2 please address the errors reported by the CIs

authmillenon commented on sys/net/gnrc/pktbuf_static/gnrc_pktbuf_static.c:
```diff
@@ -526,4 +531,54 @@ gnrc_pktsnip_t *gnrc_pktbuf_replace_snip(gnrc_pktsnip_t *pkt, gnrc_pktsnip_t *ol
+    DEBUG("ipv6_ext: duplicating %d octets\n", (int) size);
+
+    gnrc_pktsnip_t *tmp;
+    gnrc_pktsnip_t *target = gnrc_pktsnip_search_type(pkt, GNRC_NETTYPE_IPV6);
```
Using `GNRC_NETTYPE_IPV6` it doesn't make sense here...

Yonezawa-T2 replied:
Fixed to `type`.

Yonezawa-T2 commented:
If debug is disabled, the unittest now checks only the reference counter. CI should succeeds.

authmillenon commented:
Re-ACK, please squash.

Yonezawa-T2 commented:
Squashed.

cgundogan commented:
awesome!

authmillenon commented:
Thanks for the contribution by the way @Yonezawa-T2 :)
